Kubernetes 1.30 release blog (#45691)
kubernetes/website: content/en/blog/_posts/2024-04-17-kubernetes-1.30.md

**Editors:** Amit Dsouza, Frederick Kautz, Kristin Martin, Abigail McCarthy, Natali Vlatko

Announcing the release of Kubernetes v1.30: Theme TBA, the first release of 2024!

Similar to previous releases, the release of Kubernetes v1.30 introduces new stable, beta, and alpha
features. The consistent delivery of top-notch releases underscores the strength of our development
cycle and the vibrant support from our community.

This release consists of XX enhancements. Of those enhancements, XX have graduated to Stable, XX are
entering Beta, and XX have graduated to Alpha.

## Release theme and logo
<Logo image size is recommended to be no more than 2160px> 
[Kat requested to hold review for this section until after the theme is revealed, and other 
sections are reviewed.]

## Improvements that graduated to stable in Kubernetes v1.30 {#graduations-to-stable}

_This is a selection of some of the improvements that are now stable following the v1.30 release._

### Robust VolumeManager reconstruction after kubelet restart ([SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage))

Blurb for 3756 goes here.

> 

Reviewer: Suggestion:

This is a volume manager refactoring that allows the kubelet to populate additional information about how existing volumes are mounted during the kubelet startup. In general, this makes volume cleanup after kubelet restart or machine reboot more robust.

This does not bring any changes for user or cluster administrators. We used the feature process and feature gate `NewVolumeManagerReconstruction` to be able to fall back to the previous behavior in case something goes wrong. Now that the feature is GA, the feature gate is locked and cannot be disabled.

Reviewer 2: thanks @jsafrane! i've added this into the PR.

### Prevent unauthorised volume mode conversion during volume restore ([SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage))

Blurb for 3141 goes here.

### Pod Scheduling Readiness ([SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling))

Blurb for 3521 goes here.

### Min domains in PodTopologySpread ([SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling))

Blurb for 3022 goes here.

### Go workspaces for k/k ([SIG Architecture](https://github.com/kubernetes/community/tree/master/sig-architecture))

Blurb for 4402 goes here.

## Improvements that graduated to beta in Kubernetes v1.30 {#graduations-to-beta}

_This is a selection of some of the improvements that are now beta following the v1.30 release._

### Node log query ([SIG Windows](https://github.com/kubernetes/community/tree/master/sig-windows))

> Blurb for 2258 goes here.

Reviewer 3: ```suggestion
To help with debugging issues on nodes, Kubernetes v1.27 introduced a feature that allows viewing logs of services running on the node. To use the feature, ensure that the NodeLogQuery feature gate is enabled for that node, and that the kubelet configuration options enableSystemLogHandler and enableSystemLogQuery are both set to true. 

On Linux the assumption is that service logs are available via journald. On Windows the assumption is that service logs are available in the application log provider. On both operating systems, logs are also available by reading files within /var/log/.
[To read more about the feature, please click here](https://kubernetes.io/docs/concepts/cluster-administration/system-logs/#log-query)
```

### CRD Validation Ratcheting ([SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery))

Blurb for 4008 goes here--subject to Docs Freeze

### Contextual logging ([SIG Instrumentation](https://github.com/kubernetes/community/tree/master/sig-instrumentation))

Blurb for 3077 goes here.

## New alpha features

### Speed up recursive SELinux label change ([SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage))

> Blurb for 1710 goes here.

Reviewer 4: @jsafrane Looks like this one also needs a short paragraph:).

Reviewer: This is a niche and complex feature. My attempt:

```suggestion
This feature allows kubelet to apply SELinux label to volumes used in a pod in constant time using a mount option,  which speeds up pod startup on nodes with SELinux enabled. Without this feature, the container runtime had to recursively walk through the whole volumes and apply SELinux label individually to each file, which can be slow for volumes with large amount of files and directories.

Kubernetes 1.27 graduated this feature as beta, but limited it to ReadWriteOncePod volumes. The corresponding feature gate is `SELinuxMountReadWriteOncePod`. It remains beta in 1.30.

Kubernetes 1.30 extends support for SELinux mount option to all volumes as alpha, with a separate feature gate `SELinuxMount`. This feature gate introduces a behavioral change when multiple Pods with different SELinux labels share the same volume. See [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling#behavioral-changes) for details. We strongly encourage users that run  Kubernetes with SELinux enabled to test this feature and provide any feedback on the KEP.

| Feature gate | Stage in v1.30 | Behavior change |
|---|---|---| 
| SELinuxMountReadWriteOncePod |  Beta | No |
| SELinuxMount | Alpha | Yes | 

Both feature gates `SELinuxMountReadWriteOncePod` and `SELinuxMount` must be enabled to test this feature on all volumes.

This feature has no effect on Windows nodes or on Linux nodes without SELinux support.

[Kat requested to hold review for this section until after the theme is revealed, and other 
sections are reviewed.]

## Improvements that graduated to stable in Kubernetes v1.30 {#graduations-to-stable}

_This is a selection of some of the improvements that are now stable following the v1.30 release._

### Robust VolumeManager reconstruction after kubelet restart ([SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage))

This is a volume manager refactoring that allows the kubelet to populate additional information about how existing volumes are mounted during the kubelet startup. In general, this makes volume cleanup after kubelet restart or machine reboot more robust.

This does not bring any changes for user or cluster administrators. We used the feature process and feature gate `NewVolumeManagerReconstruction` to be able to fall back to the previous behavior in case something goes wrong. Now that the feature is stable, the feature gate is locked and cannot be disabled.

### Prevent unauthorised volume mode conversion during volume restore ([SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage))

> Blurb for 3141 goes here.

Reviewer 4: @RaunakShah Can you add a short paragraph here?

### Pod Scheduling Readiness ([SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling))

Blurb for 3521 goes here.

### Min domains in PodTopologySpread ([SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling))

Blurb for 3022 goes here.

### Go workspaces for k/k ([SIG Architecture](https://github.com/kubernetes/community/tree/master/sig-architecture))

Blurb for 4402 goes here.

## Improvements that graduated to beta in Kubernetes v1.30 {#graduations-to-beta}

_This is a selection of some of the improvements that are now beta following the v1.30 release._

### Node log query ([SIG Windows](https://github.com/kubernetes/community/tree/master/sig-windows))

To help with debugging issues on nodes, Kubernetes v1.27 introduced a feature that allows viewing logs of services running on the node. To use the feature, ensure that the NodeLogQuery feature gate is enabled for that node, and that the kubelet configuration options enableSystemLogHandler and enableSystemLogQuery are both set to true. 

> On Linux the assumption is that service logs are available via journald. On Windows the assumption is that service logs are available in the application log provider. On both operating systems, logs are also available by reading files within /var/log/.

Author: @Checksumz Would you please address Tim's comment?

> On Linux the assumption is that service logs are available via journald. On Windows the assumption is that service logs are available in the application log provider. On both operating systems, logs are also available by reading files within /var/log/.

Reviewer 3: @aravindhp please can you clarify the above

> On Linux the assumption is that service logs are available via journald. On Windows the assumption is that service logs are available in the application log provider. On both operating systems, logs are also available by reading files within /var/log/.

Reviewer 5: On Windows it is `C:\var\log\` which has been updated in https://github.com/kubernetes/website/pull/45691/commits/b1cd5aad95331ebcd7b9261b404b99ad6148a498

> [To read more about the feature, please click here](https://kubernetes.io/docs/concepts/cluster-administration/system-logs/#log-query)

Reviewer 6: Consider using site-relative link.

```suggestion
[To read more about the feature, please click here](/docs/concepts/cluster-administration/system-logs/#log-query)
```

Reviewer 7: We also try to avoid “click here” (instead, use the section title as the anchor)

Author: > We also try to avoid “click here” (instead, use the section title as the anchor)

Would you please add an updated suggestion for the link / link text?

Reviewer 3: @dipesh-rawat  @sftim  i have updated the title to incorporate both your changes.

### CRD Validation Ratcheting ([SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery))

Blurb for 4008 goes here--subject to Docs Freeze

### Contextual logging ([SIG Instrumentation](https://github.com/kubernetes/community/tree/master/sig-instrumentation))

Blurb for 3077 goes here.

## New alpha features

### Speed up recursive SELinux label change ([SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage))

Blurb for 1710 goes here.

### Recursive Read-only (RRO) mounts ([SIG Node](https://github.com/kubernetes/community/tree/master/sig-node))

Blurb for 3857 goes here--subject to Docs Freeze

### Job success/completion policy ([SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps))

> Blurb for 3998 goes here--subject to Docs Freeze

Reviewer 8: @natalisucks, Did you mean to extend here?

Reviewer 9: yes please @tenzen-y!

## Graduations, deprecations and removals for Kubernetes v1.30

### Graduated to stable

This lists all the features that graduated to stable (also known as _general availability_). For a
full list of updates including new features and graduations from alpha to beta, see the [release
notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.30.md).

This release includes a total of XX enhancements promoted to Stable:

(List goes here)

### Deprecations and removals
(We may not include this section, re-check if any deprecations)

### Release Notes

Check out the full details of the Kubernetes 1.30 release in our [release
notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.30.md).

### Availability

Kubernetes 1.30 is available for download on
[GitHub](https://github.com/kubernetes/kubernetes/releases/tag/v1.30.0). To get started with

> Kubernetes, check out these [interactive tutorials](https://kubernetes.io/docs/tutorials/) or run

Reviewer 6: Consider using site-relative link.

```suggestion
Kubernetes, check out these [interactive tutorials](/docs/tutorials/) or run
```

local Kubernetes clusters using [minikube](https://minikube.sigs.k8s.io/). You can also easily

> install 1.30 using [kubeadm](https://kubernetes.io/docs/setup/independent/create-cluster-kubeadm/). 

Reviewer 6: Consider using site-relative link.
```suggestion
install 1.30 using [kubeadm](/docs/setup/independent/create-cluster-kubeadm/). 
```

### Event Update

<THERE WILL BE ALWAYS A KUBECON/CLOUDNATIVECON AROUND, GIVE THE LATEST INFORMATION>

### Upcoming Release Webinar

<RELEASE WEBINARE WILL TAKE PLACE NORMALLY 30 DAYS AFTER RELEASE, ALIGN WITH CNCF TO HIGHLIGHT THE
WEBINAR>

Join members of the Kubernetes v1.30 release team on DATE AND TIME TBA to learn about the major
features of this release, as well as deprecations and removals to help plan for upgrades. For more
information and registration, visit the event page on the CNCF Online Programs site.

### Get Involved
<THIS COMMUNITY LIVES BY ITS GREAT COMMUNITY, GET THEM INVOLVED!> The simplest way to get involved
with Kubernetes is by joining one of the many [Special Interest
Groups](https://github.com/kubernetes/community/blob/master/sig-list.md) (SIGs) that align with your
interests. Have something you’d like to broadcast to the Kubernetes community? Share your voice at
our weekly [community meeting](https://github.com/kubernetes/community/tree/master/communication),
and through the channels below. Thank you for your continued feedback and support.

> - Follow us on Twitter [@Kubernetesio](https://twitter.com/kubernetesio) for latest updates

Reviewer 6: 💭 Could we consider updating the reference to "𝕏" , the current rebranded name of Twitter?

```suggestion
- Follow us on 𝕏 [@Kubernetesio](https://twitter.com/kubernetesio) for latest updates
```

Blurb for 3521 goes here.

### Min domains in PodTopologySpread ([SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling))

Blurb for 3022 goes here.

### Go workspaces for k/k ([SIG Architecture](https://github.com/kubernetes/community/tree/master/sig-architecture))

The Kubernetes repo now uses Go workspaces. This should not impact end users at all, but does have a impact for developers of downstream projects. Switching to workspaces caused some breaking changes in the flags to the various [k8s.io/code-generator](https://github.com/kubernetes/code-generator) tools. Downstream consumers should look at [`staging/src/k8s.io/code-generator/kube_codegen.sh`](https://github.com/kubernetes/code-generator/blob/master/kube_codegen.sh) to see the changes.

For full details on the changes and reasons why Go workspaces was introduced, read [Using Go workspaces in Kubernetes](https://www.kubernetes.dev/blog/2024/03/19/go-workspaces-in-kubernetes/).

## Improvements that graduated to beta in Kubernetes v1.30 {#graduations-to-beta}

_This is a selection of some of the improvements that are now beta following the v1.30 release._

> ### Node log query ([SIG Windows](https://github.com/kubernetes/community/tree/master/sig-windows))

Reviewer 3: ```suggestion
### [Node log query] (/docs/concepts/cluster-administration/system-logs/#log-query)([SIG Windows](https://github.com/kubernetes/community/tree/master/sig-windows))
```
Based on @sftim & @dipesh-rawat  suggestions

Author: I'm going to go with Abbie link suggestion from our Slack conversation

To help with debugging issues on nodes, Kubernetes v1.27 introduced a feature that allows viewing logs of services running on the node. To use the feature, ensure that the NodeLogQuery feature gate is enabled for that node, and that the kubelet configuration options enableSystemLogHandler and enableSystemLogQuery are both set to true. 

On Linux the assumption is that service logs are available via journald. On Windows the assumption is that service logs are available in the application log provider. On both operating systems, logs are also available by reading files within /var/log/.
[To read more about the feature, please click here](https://kubernetes.io/docs/concepts/cluster-administration/system-logs/#log-query)

### CRD Validation Ratcheting ([SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery))

> Blurb for 4008 goes here--subject to Docs Freeze

Reviewer 3: ```suggestion
You need to enable the `CRDValidationRatcheting`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) to
use this behavior, which then applies to all CustomResourceDefinitions in your
cluster.

Provided you enabled the feature gate, Kubernetes implements _validation racheting_
for CustomResourceDefinitions. The API server is willing accept updates to resources that
are not valid after the update, provided that each part of the resource that failed to validate
was not changed by the update operation. In other words, any invalid part of the resource
that remains invalid must have already been wrong.
You cannot use this mechanism to update a valid resource so that it becomes invalid.

This feature allows authors of CRDs to confidently add new validations to the
OpenAPIV3 schema under certain conditions. Users can update to the new schema
safely without bumping the version of the object or breaking workflows.
```

forged over several release cycles. A very special thanks goes out our release lead, Kat Cosgrove, 
for supporting us through a successful release cycle, advocating for us, making sure that we could
all contribute in the best way possible, and challenging us to improve the release process.

### Project Velocity

The CNCF K8s DevStats project aggregates a number of interesting data points related to the velocity of Kubernetes and various sub-projects. This includes everything from individual contributions to the number of companies that are contributing and is an illustration of the depth and breadth of effort that goes into evolving this ecosystem.

In the v1.30 release cycle, which ran for 14 weeks (January 8 to April 17), we saw contributions from [888 companies](https://k8s.devstats.cncf.io/d/9/companies-table?orgId=1&var-period_name=v1.29.0%20-%20now&var-metric=contributions) and [1268 individuals](https://k8s.devstats.cncf.io/d/66/developer-activity-counts-by-companies?orgId=1&var-period_name=v1.29.0%20-%20now&var-metric=contributions&var-repogroup_name=Kubernetes&var-repo_name=kubernetes%2Fkubernetes&var-country_name=All&var-companies=All).

### Event Update

> <THERE WILL BE ALWAYS A KUBECON/CLOUDNATIVECON AROUND, GIVE THE LATEST INFORMATION>

Reviewer 3: ```suggestion
* KubeCon + CloudNativeCon China 2024 will take place in Hong Kong, from 21 – 23 August 2024! You can find more information about the conference and registration on the [event site](https://events.linuxfoundation.org/kubecon-cloudnativecon-open-source-summit-ai-dev-china/).
* KubeCon + CloudNativeCon North America 2024 will take place in Salt Lake City, Utah, The United States of America, from 12 – 15 November 2024! You can find more information about the conference and registration on the [event site](https://events.linuxfoundation.org/kubecon-cloudnativecon-north-america/).
```

### Upcoming Release Webinar

> <RELEASE WEBINARE WILL TAKE PLACE NORMALLY 30 DAYS AFTER RELEASE, ALIGN WITH CNCF TO HIGHLIGHT THE
> WEBINAR>

Reviewer 3: ```suggestion
Join members of the Kubernetes v1.30 release team on Thursday, May 23rd, 2024, at 9 A.M. PT to learn about the major features of this release, as well as deprecations and removals to help plan for upgrades. For more information and registration, visit the [event page](https://community.cncf.io/events/details/cncf-cncf-online-programs-presents-cncf-live-webinar-kubernetes-130-release/) on the CNCF Online Programs site.
```

---
layout: blog
title: 'Kubernetes v1.30: TBA'
date: 2024-04-17
slug: kubernetes-v1-30-release
---

**Authors:** [Kubernetes v1.30 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.30/release_team.md)

> **Editors:** Amit Dsouza, Frederick Kautz, Kristin Martin, Abigail McCarthy, Natali Vlatko

Reviewer 7: (aside) This could be hard to localize. How do we explain the difference between the authors of the article and the editors of the article, so that localization teams can pick the right word?

Author: We did the same with the [1.29 Release Blog](https://kubernetes.io/blog/2023/12/13/kubernetes-v1-29-release/). I'll check if there were any localization issues with that one.

Author: No issues noted by localization team member with the previous blog that used the same byline format

Announcing the release of Kubernetes v1.30: Uwubernetes, the cutest release!

Similar to previous releases, the release of Kubernetes v1.30 introduces new stable, beta, and alpha
features. The consistent delivery of top-notch releases underscores the strength of our development
cycle and the vibrant support from our community.

This release consists of 45 enhancements. Of those enhancements, 17 have graduated to Stable, 18 are
entering Beta, and 10 have graduated to Alpha.

## Release theme and logo

> (logo placeholder-pending high res file and resize)

Reviewer 7: (TBC)

Kubernetes v1.30 makes your clusters cuter!

Kubernetes is built and released by thousands of people from all over the world and all walks of
life. Most contributors are not being paid to do this; we build it for fun, to solve a problem, to
learn something, or for the simple love of the community. Many of us found our homes, our friends,
and our careers here. The Release Team is honored to be a part of the continued growth of
Kubernetes. 

For the people who built it, for the people who release it, and for the furries who keep all of our
clusters online, we present to you Kubernetes v1.30: Uwubernetes, the cutest release to date. The
name is a portmanteau of “kubernetes” and “UwU,” an emoticon used to indicate happiness or cuteness.
We’ve found joy here, but we’ve also brought joy from our outside lives that helps to make this
community as weird and wonderful and welcoming as it is. We’re so happy to share our work with you. 

> UwU :heart:

Reviewer 7: ```suggestion
UwU ♥️
```

(unless we want the literal text `:heart:`)

Author: Kat's intro shows a heart emoji, this [chart](https://gist.github.com/rxaviers/7360908) shows `:heart:` to be the correct markdown entity.

Author: Okay, it seems GH supports emojis, while our site doesn't. Will change to unicode.

Reviewer 7: GitHub uses [GitHub flavored Markdown](https://github.github.com/gfm/); we use [CommonMark](https://commonmark.org/) (with a few Hugo extensions).

## Improvements that graduated to stable in Kubernetes v1.30 {#graduations-to-stable}

_This is a selection of some of the improvements that are now stable following the v1.30 release._

### Robust VolumeManager reconstruction after kubelet restart ([SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage))

This is a volume manager refactoring that allows the kubelet to populate additional information about how existing volumes are mounted during the kubelet startup. In general, this makes volume cleanup after kubelet restart or machine reboot more robust.

This does not bring any changes for user or cluster administrators. We used the feature process and feature gate `NewVolumeManagerReconstruction` to be able to fall back to the previous behavior in case something goes wrong. Now that the feature is stable, the feature gate is locked and cannot be disabled.

### Prevent unauthorized volume mode conversion during volume restore ([SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage))

> With the release of Kubernetes 1.30, the ability to inherently prevent unauthorized users from converting the volume mode of a PVC, when it is created from an existing VolumeSnapshot, moves to GA. Cluster administrators will now have to specifically grant permissions in Kubernetes to authorized users, like storage backup vendors, to allow them to perform such actions. 

Reviewer 7: Try this:
```suggestion
For Kubernetes 1.30, the control plane always prevents unauthorized changes to volume modes when restoring
a snapshot into a PersistentVolume. As a cluster administrator, you'll need to grant permissions to the appropriate identity principals (for example: ServiceAccounts representing a storage integration) if you
need to allow that kind of change at restore time.

```

For more information, see [converting the volume mode of a Snapshot](/docs/concepts/storage/volume-snapshots/#convert-volume-mode).

### Pod Scheduling Readiness ([SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling))

> Blurb for 3521 goes here.

Reviewer 9: ```suggestion
Pod Scheduling Readiness introduces a new phase at the start of a Pod's lifecycle. A Pod won't be scheduled until its `schedulerGates` are all cleared. This feature empowers external integrators to intercede during the pre-scheduling cycle of a Pod, thereby enabling the orchestration of customized workflows that cater to specific in-house requirements.
```

Reviewer 7: Rather than talking about “introducing a new phase”, how about explaining that the Pod lifecycle
has a phase where the scheduler checks if the Pod has scheduling suspended?

Author: @natalisucks ptal

Reviewer 9: @sftim I have the understanding that Pods are ready to be scheduled once they're created, and the `schedulingGates` field can only be initialized when the Pod is created. With that being the case, could we possibly restate that first sentence as "This feature enhances the Pod lifecycle by allowing you to control when a Pod can be scheduled, downsizing the churn that the Kubernetes scheduler might encounter."

Reviewer 7: > @sftim I have the understanding that Pods are ready to be scheduled once they're created

Sometimes. Pods are ready to be scheduled if either: they are created without scheduling gates, or if they were created with scheduling gates that have since been removed.

---

I'm not particularly keen on “this feature&hellip;” or “enhances” for _graduated_ features. At graduation, we're describing the stable behavior of Kubernetes. We're also discussing something that's been on by default for at least one minor release and possibly many. Yes, it absolutely is an enhancement, but to a typical end user the change is around the level of support, not actually the behavior.

Reviewer 9: @sftim Do you have a suggestion to replace "feature" as the descriptor of this enhancement as opposed to repeating the name of it? Or would you be more keen on repeating the name of the enhancement?

### Min domains in PodTopologySpread ([SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling))

> Blurb for 3022 goes here.

Reviewer 9: ```suggestion
The `minDomains` parameter for PodTopologySpread constraints graduates to stable this release, which allows you to define the minimum number of domains. This feature is designed to be used with the cluster autoscaler.

If you previously attempted use and there weren't enough domains already present, Pods would be marked as unschedulable. The cluster autoscaler would then provision Node(s) in new domain(s), and you'd eventually get Pods spreading over enough domains.
```

For full details on the changes and reasons why Go workspaces was introduced, read [Using Go workspaces in Kubernetes](https://www.kubernetes.dev/blog/2024/03/19/go-workspaces-in-kubernetes/).

## Improvements that graduated to beta in Kubernetes v1.30 {#graduations-to-beta}

_This is a selection of some of the improvements that are now beta following the v1.30 release._

### Node log query ([SIG Windows](https://github.com/kubernetes/community/tree/master/sig-windows))

To help with debugging issues on nodes, Kubernetes v1.27 introduced a feature that allows viewing logs of services running on the node. To use the feature, ensure that the NodeLogQuery feature gate is enabled for that node, and that the kubelet configuration options enableSystemLogHandler and enableSystemLogQuery are both set to true. 

On Linux the assumption is that service logs are available via journald. On Windows the assumption is that service logs are available in the application log provider. On both operating systems, logs are also available by reading files within /var/log/.
For more information, see the [log query](/docs/concepts/cluster-administration/system-logs/#log-query) documentation.

> ### CRD Validation Ratcheting ([SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery))

Reviewer 7: Try:
```suggestion
### CRD validation ratcheting ([SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery))
```

You need to enable the `CRDValidationRatcheting`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) to
use this behavior, which then applies to all CustomResourceDefinitions in your
cluster.

Provided you enabled the feature gate, Kubernetes implements _validation racheting_
for CustomResourceDefinitions. The API server is willing accept updates to resources that
are not valid after the update, provided that each part of the resource that failed to validate
was not changed by the update operation. In other words, any invalid part of the resource
that remains invalid must have already been wrong.
You cannot use this mechanism to update a valid resource so that it becomes invalid.

This feature allows authors of CRDs to confidently add new validations to the
OpenAPIV3 schema under certain conditions. Users can update to the new schema
safely without bumping the version of the object or breaking workflows.

### Contextual logging ([SIG Instrumentation](https://github.com/kubernetes/community/tree/master/sig-instrumentation))

> Blurb for 3077 goes here.

Reviewer 10: ```suggestion
Contextual Logging advances to beta in Kubernetes 1.30, empowering developers and operators to inject customizable, correlatable contextual details like service names and transaction IDs into logs through `WithValues` and `WithName`. This enhancement simplifies the correlation and analysis of log data across distributed systems, significantly improving the efficiency of troubleshooting efforts. By offering a clearer insight into the workings of your Kubernetes environments, Contextual Logging ensures that operational challenges are more manageable, marking a notable step forward in Kubernetes observability.
```

Reviewer 7: OK, but: please wrap the Markdown at around 100 characters.

Reviewer 9: @sftim Kristin will do an overall pass to wrap all lines as discussed in our release comms meeting recently

### Make Kubernetes aware of the LoadBalancer behaviour ([SIG Network](https://github.com/kubernetes/community/tree/master/sig-network))

Blurb for 1860 goes here.

## New alpha features

### Speed up recursive SELinux label change ([SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage))

> This feature allows kubelet to apply SELinux label to volumes used in a pod in constant time using a mount option,  which speeds up pod startup on nodes with SELinux enabled. Without this feature, the container runtime had to recursively walk through the whole volumes and apply SELinux label individually to each file, which can be slow for volumes with large amount of files and directories.

Reviewer 7: How about:
```suggestion
From the v1.27 release, Kubernetes already included an optimization that sets SELinux labels
on the contents of volumes, using only constant time.
Kubernetes achieves that speed up using a mount option. The slower legacy behavior requires
the container runtime to recursively walk through the whole volumes and apply SELinux labelling
individually to each file and directory; this is especially noticable for volumes with large amount of
files and directories.
```

> Kubernetes 1.27 graduated this feature as beta, but limited it to ReadWriteOncePod volumes. The corresponding feature gate is `SELinuxMountReadWriteOncePod`. It remains beta in 1.30.

Reviewer 7: ```suggestion
Kubernetes 1.27 graduated this feature as beta, but limited it to ReadWriteOncePod volumes.
The corresponding feature gate is `SELinuxMountReadWriteOncePod`. It's still enabled by default
and remains beta in 1.30.
```

> Kubernetes 1.30 extends support for SELinux mount option to all volumes as alpha, with a separate feature gate `SELinuxMount`. This feature gate introduces a behavioral change when multiple Pods with different SELinux labels share the same volume. See [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling#behavioral-changes) for details. We strongly encourage users that run  Kubernetes with SELinux enabled to test this feature and provide any feedback on the KEP.

Reviewer 7: ```suggestion
Kubernetes 1.30 extends support for SELinux mount option to **all** volumes as alpha,
with a separate feature gate: `SELinuxMount`. This feature gate introduces a behavioral change
when multiple Pods with different SELinux labels share the same volume.
See [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling#behavioral-changes) for details.
We strongly encourage users that run Kubernetes with SELinux enabled to test this feature and
provide any feedback on the [KEP issue](https://kep.k8s.io/1710).
```

| Feature gate | Stage in v1.30 | Behavior change |
|---|---|---| 
| SELinuxMountReadWriteOncePod |  Beta | No |
| SELinuxMount | Alpha | Yes | 

Both feature gates `SELinuxMountReadWriteOncePod` and `SELinuxMount` must be enabled to test this feature on all volumes.

This feature has no effect on Windows nodes or on Linux nodes without SELinux support.

### Recursive Read-only (RRO) mounts ([SIG Node](https://github.com/kubernetes/community/tree/master/sig-node))

> Blurb for 3857 goes here--subject to Docs Freeze

Reviewer 7: (FIXME)

Author: @fkautz please add this section

Reviewer 10: ```suggestion
Introducing Recursive Read-Only (RRO) Mounts in alpha this release, you'll find a new layer of security for your data. This feature lets you set volumes and their submounts as read-only, preventing accidental modifications. Imagine deploying a critical application where data integrity is key—RRO Mounts ensure that your data stays untouched, reinforcing your cluster's security with an extra safeguard. This is especially crucial in tightly controlled environments, where even the slightest change can have significant implications.
```

### Job success/completion policy ([SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps))

Blurb for 3998 goes here--subject to Docs Freeze

### Traffic distribution for services ([SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps))

> Blurb for 4444 goes here

Reviewer 7: (FIXME)

## Graduations, deprecations and removals for Kubernetes v1.30

### Graduated to stable

This lists all the features that graduated to stable (also known as _general availability_). For a
full list of updates including new features and graduations from alpha to beta, see the [release
notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.30.md).

> This release includes a total of XX enhancements promoted to Stable:

Reviewer 9: ```suggestion
This release includes a total of 16 enhancements promoted to stable:
```

Reviewer 9: @kcmartin would love for you to double-check this number just in case I've miscounted!

Author: I see 17 on the project board. We'll do a final stats check for all the numbers we cite in the blog.

> (List goes here)

Reviewer 9: ```suggestion
- [Container Resource based Pod Autoscaling](https://kep.k8s.io/1610)
- [Remove transient node predicates from KCCM's service controller](https://kep.k8s.io/3458)
- [Go workspaces for k/k](https://kep.k8s.io/4402)
- [Reduction of Secret-based Service Account Tokens](https://kep.k8s.io/2799)
- [CEL for Admission Control](https://kep.k8s.io/3488)
- [CEL-based admission webhook match conditions](https://kep.k8s.io/3716)
- [Pod Scheduling Readiness](https://kep.k8s.io/3521)
- [Min domains in PodTopologySpread](https://kep.k8s.io/3022)
- [Prevent unauthorised volume mode conversion during volume restore](https://kep.k8s.io/3141)
- [API Server Tracing](https://kep.k8s.io/647)
- [Cloud Dual-Stack --node-ip Handling](https://kep.k8s.io/3705)
- [AppArmor support](https://kep.k8s.io/24)
- [Robust VolumeManager reconstruction after kubelet restart](https://kep.k8s.io/3756)
- [kubectl delete: Add interactive(-i) flag](https://kep.k8s.io/3895)
- [Metric cardinality enforcement](https://kep.k8s.io/2305)
- [Field `status.hostIPs` added for Pod](https://kep.k8s.io/2681)
```

### Deprecations and removals

> (We may not include this section, re-check if any deprecations)

Reviewer 9: ```suggestion
### Removed the SecurityContextDeny admission plugin, deprecated since v1.27 ([SIG Auth](https://github.com/kubernetes/community/tree/master/sig-auth), [SIG Security](https://github.com/kubernetes/community/tree/master/sig-security), and [SIG Testing](https://github.com/kubernetes/community/tree/master/sig-testing))

With the removal of the SecurityContextDeny admission plugin, the Pod Security Admission plugin, available since v1.25, is recommended instead. See: [SecurityContextDeny reference documentation](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#securitycontextdeny) for more information.
```

Reviewer 7: - We might want to simplify the fragment identifier here.
- Relative hyperlink please
- Check that the v1.30 docs include the thing we're linking to; after all, we're removing it.

Reviewer 7: - We might want to simplify the fragment identifier here.
- Relative hyperlink please
- Check that the v1.30 docs include the thing we're linking to; after all, we're removing it.

Reviewer 9: @sftim this is the removal in question: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.30.md#deprecation

Reviewer 7: Ah, so the changelog needs fixing as well (this might be my text, even :grimacing:). The problem is that we've remove the documentation, so that link won't work after the release.

I [commented](https://github.com/kubernetes/kubernetes/pull/122612#issuecomment-2030080190) to that effect.

Author: Is there any replacement documentation link, or are we omitting the link?

Reviewer 7: My advice, similar to the changelog suggestion: omit the link. It's not a feature people were actually using.

> ### Release Notes

Reviewer 7: ```suggestion
## Release notes
```

Check out the full details of the Kubernetes 1.30 release in our [release
notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.30.md).

> ### Availability

Reviewer 7: ```suggestion
## Availability
```

Kubernetes 1.30 is available for download on
[GitHub](https://github.com/kubernetes/kubernetes/releases/tag/v1.30.0). To get started with
Kubernetes, check out these [interactive tutorials](/docs/tutorials/) or run
local Kubernetes clusters using [minikube](https://minikube.sigs.k8s.io/). You can also easily
install 1.30 using [kubeadm](/docs/setup/independent/create-cluster-kubeadm/). 

> ### Release Team

Reviewer 7: ```suggestion
## Release team
```

Kubernetes is only possible with the support, commitment, and hard work of its community. Each
release team is made up of dedicated community volunteers who work together to build the many pieces
that make up the Kubernetes releases you rely on. This requires the specialized skills of people
from all corners of our community, from the code itself to its documentation and project management.

We would like to thank the entire [release team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.30/release_team.md)
for the hours spent hard at work to deliver the Kubernetes v1.30 release to our community. The 
Release Team's membership ranges from first-time shadows to returning team leads with experience 
forged over several release cycles. A very special thanks goes out our release lead, Kat Cosgrove, 
for supporting us through a successful release cycle, advocating for us, making sure that we could
all contribute in the best way possible, and challenging us to improve the release process.

> ### Project Velocity

Reviewer 7: ```suggestion
## Project velocity
```

The CNCF K8s DevStats project aggregates a number of interesting data points related to the velocity of Kubernetes and various sub-projects. This includes everything from individual contributions to the number of companies that are contributing and is an illustration of the depth and breadth of effort that goes into evolving this ecosystem.

In the v1.30 release cycle, which ran for 14 weeks (January 8 to April 17), we saw contributions from [888 companies](https://k8s.devstats.cncf.io/d/9/companies-table?orgId=1&var-period_name=v1.29.0%20-%20now&var-metric=contributions) and [1268 individuals](https://k8s.devstats.cncf.io/d/66/developer-activity-counts-by-companies?orgId=1&var-period_name=v1.29.0%20-%20now&var-metric=contributions&var-repogroup_name=Kubernetes&var-repo_name=kubernetes%2Fkubernetes&var-country_name=All&var-companies=All).

> ### Event Update

Reviewer 7: ```suggestion
## Event update
```

<THERE WILL BE ALWAYS A KUBECON/CLOUDNATIVECON AROUND, GIVE THE LATEST INFORMATION>

> ### Upcoming Release Webinar

Reviewer 7: ```suggestion
## Upcoming release webinar
```

<RELEASE WEBINARE WILL TAKE PLACE NORMALLY 30 DAYS AFTER RELEASE, ALIGN WITH CNCF TO HIGHLIGHT THE
WEBINAR>

Join members of the Kubernetes v1.30 release team on DATE AND TIME TBA to learn about the major
features of this release, as well as deprecations and removals to help plan for upgrades. For more
information and registration, visit the event page on the CNCF Online Programs site.

> ### Get Involved
> <THIS COMMUNITY LIVES BY ITS GREAT COMMUNITY, GET THEM INVOLVED!> The simplest way to get involved

Reviewer 7: ```suggestion
## Get involved

The simplest way to get involved
```

This is a volume manager refactoring that allows the kubelet to populate additional information about how existing volumes are mounted during the kubelet startup. In general, this makes volume cleanup after kubelet restart or machine reboot more robust.

This does not bring any changes for user or cluster administrators. We used the feature process and feature gate `NewVolumeManagerReconstruction` to be able to fall back to the previous behavior in case something goes wrong. Now that the feature is stable, the feature gate is locked and cannot be disabled.

### Prevent unauthorized volume mode conversion during volume restore ([SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage))

For Kubernetes 1.30, the control plane always prevents unauthorized changes to volume modes when restoring
a snapshot into a PersistentVolume. As a cluster administrator, you'll need to grant permissions to the appropriate identity principals (for example: ServiceAccounts representing a storage integration) if you
need to allow that kind of change at restore time.

> For more information, see [converting the volume mode of a Snapshot](/docs/concepts/storage/volume-snapshots/#convert-volume-mode).

Reviewer 2: @xing-yang I've added in a call out to highlight the upgrade notes. let me know if you think its enough

> For more information, see [converting the volume mode of a Snapshot](/docs/concepts/storage/volume-snapshots/#convert-volume-mode).

Reviewer 4: Looks good!  Thanks.

| Feature gate | Stage in v1.30 | Behavior change |
|---|---|---| 
| SELinuxMountReadWriteOncePod |  Beta | No |
| SELinuxMount | Alpha | Yes | 

Both feature gates `SELinuxMountReadWriteOncePod` and `SELinuxMount` must be enabled to test this feature on all volumes.

This feature has no effect on Windows nodes or on Linux nodes without SELinux support.

### Recursive Read-only (RRO) mounts ([SIG Node](https://github.com/kubernetes/community/tree/master/sig-node))

Blurb for 3857 goes here--subject to Docs Freeze

### Job success/completion policy ([SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps))

Blurb for 3998 goes here--subject to Docs Freeze

### Traffic distribution for services ([SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps))

Blurb for 4444 goes here

## Graduations, deprecations and removals for Kubernetes v1.30

### Graduated to stable

This lists all the features that graduated to stable (also known as _general availability_). For a
full list of updates including new features and graduations from alpha to beta, see the [release
notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.30.md).

> This release includes a total of 17 enhancements promoted to Stable:

Reviewer 9: we've received confirmation from @drewhagen that this KEP is indeed fine and should be included – adding it below!

with a separate feature gate: `SELinuxMount`. This feature gate introduces a behavioral change
when multiple Pods with different SELinux labels share the same volume.
See [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling#behavioral-changes) for details.
We strongly encourage users that run Kubernetes with SELinux enabled to test this feature and
provide any feedback on the [KEP issue](https://kep.k8s.io/1710).

| Feature gate | Stage in v1.30 | Behavior change |
|---|---|---| 
| SELinuxMountReadWriteOncePod |  Beta | No |
| SELinuxMount | Alpha | Yes | 

Both feature gates `SELinuxMountReadWriteOncePod` and `SELinuxMount` must be enabled to test this feature on all volumes.

This feature has no effect on Windows nodes or on Linux nodes without SELinux support.

### Recursive Read-only (RRO) mounts ([SIG Node](https://github.com/kubernetes/community/tree/master/sig-node))

Blurb for 3857 goes here--subject to Docs Freeze

### Job success/completion policy ([SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps))

Blurb for 3998 goes here--subject to Docs Freeze

> ### Traffic distribution for services ([SIG Network](https://github.com/kubernetes/community/tree/master/sig-network))

Reviewer 2: @gauravkghildiyal can you take a look here?

Reviewer 11: Thanks! LGTM

Kubernetes 1.30 extends support for SELinux mount option to **all** volumes as alpha,
with a separate feature gate: `SELinuxMount`. This feature gate introduces a behavioral change
when multiple Pods with different SELinux labels share the same volume.
See [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling#behavioral-changes) for details.
We strongly encourage users that run Kubernetes with SELinux enabled to test this feature and
provide any feedback on the [KEP issue](https://kep.k8s.io/1710).

| Feature gate | Stage in v1.30 | Behavior change |
|---|---|---| 
| SELinuxMountReadWriteOncePod |  Beta | No |
| SELinuxMount | Alpha | Yes | 

Both feature gates `SELinuxMountReadWriteOncePod` and `SELinuxMount` must be enabled to test this feature on all volumes.

This feature has no effect on Windows nodes or on Linux nodes without SELinux support.

### Recursive Read-only (RRO) mounts ([SIG Node](https://github.com/kubernetes/community/tree/master/sig-node))

Blurb for 3857 goes here--subject to Docs Freeze

### Job success/completion policy ([SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps))

> Blurb for 3998 goes here--subject to Docs Freeze

Reviewer 9: ```suggestion
From Kubernetes v1.30, indexed Jobs support `.spec.successPolicy` to define when a Job can be declared succeeded based on succeeded Pods. This allows you to define two types of criteria:

- `succeededIndexes` indicates that the Job can be declared succeeded when these indexes succeeded, even if other indexes failed.
- `succeededCount` indicates that the Job can be declared succeeded when the number of succeeded Indexes reaches this criterion.

After the Job meets the success policy, the Job controller terminates the lingering Pods.
```

Reviewer 9: @kcmartin @sftim please note: Yuki (@tenzen-y) has had a problem with their GitHub account (currently being reviewed by the GitHub team), so the KEP for this feature isn't currently available. Yuki has kindly sent this blurb for the KEP via Slack, and I've done my best to grammar edit. Your review is also welcome, thanks!

notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.30.md).

This release includes a total of 17 enhancements promoted to Stable:

- [Container Resource based Pod Autoscaling](https://kep.k8s.io/1610)
- [Remove transient node predicates from KCCM's service controller](https://kep.k8s.io/3458)
- [Go workspaces for k/k](https://kep.k8s.io/4402)
- [Reduction of Secret-based Service Account Tokens](https://kep.k8s.io/2799)
- [CEL for Admission Control](https://kep.k8s.io/3488)
- [CEL-based admission webhook match conditions](https://kep.k8s.io/3716)
- [Pod Scheduling Readiness](https://kep.k8s.io/3521)
- [Min domains in PodTopologySpread](https://kep.k8s.io/3022)
- [Prevent unauthorised volume mode conversion during volume restore](https://kep.k8s.io/3141)
- [API Server Tracing](https://kep.k8s.io/647)
- [Cloud Dual-Stack --node-ip Handling](https://kep.k8s.io/3705)
- [AppArmor support](https://kep.k8s.io/24)
- [Robust VolumeManager reconstruction after kubelet restart](https://kep.k8s.io/3756)
- [kubectl delete: Add interactive(-i) flag](https://kep.k8s.io/3895)
- [Metric cardinality enforcement](https://kep.k8s.io/2305)

> - [Field `status.hostIPs` added for Pod](https://kep.k8s.io/2681)

Reviewer 9: ```suggestion
- [Field `status.hostIPs` added for Pod](https://kep.k8s.io/2681)
- [Aggregated Discovery](https://kep.k8s.io/3352)
```

For Kubernetes 1.30, the control plane always prevents unauthorized changes to volume modes when restoring
a snapshot into a PersistentVolume. As a cluster administrator, you'll need to grant permissions to the appropriate identity principals (for example: ServiceAccounts representing a storage integration) if you
need to allow that kind of change at restore time.

For more information, see [converting the volume mode of a Snapshot](/docs/concepts/storage/volume-snapshots/#convert-volume-mode).

### Pod Scheduling Readiness ([SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling))

_Pod scheduling readiness_ graduates to stable this release, after being promoted to beta in Kubernetes v1.27.

This now-stable feature lets Kubernetes avoid trying to schedule a Pod that has been defined, when the cluster
doesn't yet have the resources provisioned to allow actually binding that Pod to a node. That's not the only use
case; the custom control on whether a Pod can be allowed to schedule also lets you implement quota mechanisms,

> security controls, and more.

Reviewer 9: nit: do we want to add another line break here to match the rest of the formatting of the other sections?

---
layout: blog
title: 'Kubernetes v1.30: Uwubernetes'
date: 2024-04-17
slug: kubernetes-v1-30-release

> author: [Kubernetes v1.30 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.30/release-team.md)

Reviewer 7: ```suggestion
author: >
  [Kubernetes v1.30 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.30/release-team.md)
```

> **Authors:** [Kubernetes v1.30 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.30/release_team.md)

Reviewer 12: ```suggestion
**Authors:** [Kubernetes v1.30 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.30/release-team.md)
```
@kcmartin, I also saw the url link broken to the 1.30 release team

> ---
> 
> **Authors:** [Kubernetes v1.30 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.30/release-team.md)

Reviewer 7: ```suggestion
author: [Kubernetes v1.30 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.30/release-team.md)
---

```

> **Editors:** Amit Dsouza, Frederick Kautz, Kristin Martin, Abigail McCarthy, Natali Vlatko

Reviewer 7: Don't put `Editors: Amit Dsouza, Frederick Kautz, Kristin Martin, Abigail McCarthy, Natali Vlatko` in the `author` field. If we really want to support editors, we should add proper support for a new front matter field.

logs of services running on the node. To use the feature, ensure that the `NodeLogQuery` feature
gate is enabled for that node, and that the kubelet configuration options `enableSystemLogHandler`
and `enableSystemLogQuery` are both set to true.

Following the v1.30 release, this is now beta (you still need to enable the feature to use it,
though).

On Linux the assumption is that service logs are available via journald. On Windows the assumption
is that service logs are available in the application log provider. Logs are also available by
reading files within `/var/log/` (Linux) or `C:\var\log\` (Windows). For more information, see the
[log query](/docs/concepts/cluster-administration/system-logs/#log-query) documentation.

### CRD validation ratcheting ([SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery))

You need to enable the `CRDValidationRatcheting` [feature
gate](/docs/reference/command-line-tools-reference/feature-gates/) to use this behavior, which then
applies to all CustomResourceDefinitions in your cluster.

Provided you enabled the feature gate, Kubernetes implements _validation racheting_ for

> CustomResourceDefinitions. The API server is willing accept updates to resources that are not valid

Reviewer 13: ```suggestion
CustomResourceDefinitions. The API server is willing to accept updates to resources that are not valid
```
